hal/src/rtl872x/i2c_hal.cpp

```diff
@@ -174,29 +174,45 @@ class I2cClass {
             // RCC_PeriphClockCmd(APBPeriph_I2C0, APBPeriph_I2C0_CLOCK, DISABLE);
             Pinmux_Config(hal_pin_to_rtl_pin(sdaPin_), PINMUX_FUNCTION_GPIO);
             Pinmux_Config(hal_pin_to_rtl_pin(sclPin_), PINMUX_FUNCTION_GPIO);
+            txBuffer_.reset();
+            rxBuffer_.reset();
         }
         return SYSTEM_ERROR_NONE;
     }
 
+    int suspend() {
+        CHECK_TRUE(state_ = HAL_I2C_STATE_ENABLED, SYSTEM_ERROR_INVALID_STATE);
+        CHECK(end());
+        state_ = HAL_I2C_STATE_SUSPENDED;
+        return SYSTEM_ERROR_NONE;
+    }
+
+    int restore() {
+        CHECK_TRUE(state_ = HAL_I2C_STATE_SUSPENDED, SYSTEM_ERROR_INVALID_STATE);
+        return begin((i2cInitStruct_.I2CMaster == I2C_MASTER_MODE) ? I2C_MODE_MASTER : I2C_MODE_SLAVE, i2cInitStruct_.I2CAckAddr);
+    }
+
     void reset() {
         end();
 
+        hal_gpio_mode(sclPin_, INPUT_PULLUP);
         // Just in case make sure that the pins are correctly configured (they should anyway be at this point)
         hal_gpio_config_t conf = {};
         conf.size = sizeof(conf);
         conf.version = HAL_GPIO_VERSION;
         conf.mode = OUTPUT_OPEN_DRAIN_PULLUP;
         conf.set_value = true;
-        conf.value = 1;
-        hal_gpio_configure(sdaPin_, &conf, nullptr);
-        conf.value = hal_gpio_read(sclPin_);
+        conf.value = hal_gpio_read(sclPin_);;
         hal_gpio_configure(sclPin_, &conf, nullptr);
 
         // Generate up to 9 pulses on SCL to tell slave to release the bus
         for (int i = 0; i < 9; i++) {
+            conf.value = 1;
+            hal_gpio_configure(sdaPin_, &conf, nullptr);
             hal_gpio_write(sdaPin_, 1);
             HAL_Delay_Microseconds(50);
 
+            hal_gpio_mode(sdaPin_, INPUT_PULLUP);
```

Comment:
We have to configure it to input to get correct external signal as per the user manual: 
>When Port A is configured as Input, then reading this location reads the values on the signal. When the data direction of Port A is set as Output, reading this location reads the data register for Port A.

Reply:
Then we have a problem with our gpio hal implementation. `OUTPUT_OPEN_DRAIN_PULLUP` is supposed to read the input pad of the pin. Where is this quote coming from?

Reply:
The best fix for this is probably to use `direction` to control the state of the pin while in `OPEN_DRAIN` modes. Writing 0 should configure the direction to output and force the pin to 0. Reading the pin state (and its actual state) will give us 0. Writing 1 should switch the direction to input and in case of `OUTPUT_OPEN_DRAIN_PULLUP` we'll get idle (pulled) state when reading the pin unless something else is pulling it low, in case of `OUTPUT_OPEN_DRAIN` and proper external pull-up we'll also get idle (pulled) state when reading the pin unless something else is pulling it low.

Reply:
That is quouted from the description of the `gpio_ext_porta` register, which is supposed to reflect the signal on the external pad. Regarding your suggestion, I'm just afraid that writing 1 might not have enough drive strength to drive external component.

Reply:
There is no driving to 1 in open-drain configuration? Drive strength to 1 solely depends on the pull-up value (and pinmux controller has pull-up strength selector).

Reply:
Okay. I'll test if that works as expected by changing the GPIO HAL.

Reply:
I just notice that if a pin is configured as output open drain and output 1,  which means that in our new driver the pin is actually configured as input open drain, and there is a external pull down resistor, we will read back with 0, which  violates the expected 1.

Reply:
No, it doesn't violate, that's the correct expected behavior. Again, in open-drain you can only pull the pin down (you have high-side transistor disconnected). It can be pulled down by us, or something else (e.g. some other IC), in both cases you should be reading 0.

Reply:
Imagine this use case: a pin is used to drive a led by high level and to avoid glitch on powering up, user adds a external pull-down resistor. When user calls `pinMode(led, OUTPUT)` followed by `digitalWrite(led, HIGH)`, the led should be lighted up. But due to the actual pin direction, it won't, no?

Reply:
That's not how open-drain mode behaves. If you add external pull-down on a pin that is configured as open-drain, you cannot drive it high in any way. Again, the high side transistor is disconnected, the only way to pull it up is using a pull-up resistor. There is only low-side transistor to switch it to 0, or to disconnect low-side and return back to 1.

Reply:
I see. Then we need to configure a pin to output with pull-up when `pinMode(pin, OUTPUT)` is called to correctly drive external components.

Reply:
Why? `OUTPUT` is its own gpio mode: push-pull. It has nothing to do with open-drain configuration. Push-pull is the regular configuration with both high and low side transistor available to either bring to 1 or 0.

Reply:
Just to clarify again, the proposed fix that I explained in https://github.com/particle-iot/device-os/pull/2497#discussion_r943398385 only applies to open-drain configurations (OUTPUT_OPEN_DRAIN/OUTPUT_OPEN_DRAIN_PULLUP), not OUTPUT (push-pull).

Reply:
`OUTPUT` and `OUTPUT_OPEN_DRAIN` apply the same configuration for now.

Reply:
Yup, and that needs to change, given the limitations of the GPIO peripheral, hence my initial comment that we have an issue in GPIO HAL and we fixed the symptoms of that in this PR e.g. in the tests and not the source of the problem.